modules/integration/src/test/java/io/ballerina/messaging/broker/integration/standalone/jms/QueueDistributedTransactionTest.java

```diff
@@ -0,0 +1,248 @@
+
+        XAConnectionFactory xaConnectionFactory =
+                (XAConnectionFactory) initialContext.lookup(ClientHelper.XA_CONNECTION_FACTORY);
+
+        XAConnection xaConnection = xaConnectionFactory.createXAConnection();
+        XASession xaSession = xaConnection.createXASession();
+        XAResource xaResource = xaSession.getXAResource();
+
+        Session session = xaSession.getSession();
+        Queue queue = session.createQueue(queueName);
+        MessageProducer producer = session.createProducer(queue);
+        xaConnection.start();
+
+        XidImpl xid = new XidImpl(0, "branchId_1".getBytes(), "globalId_1".getBytes());
+        xaResource.start(xid, XAResource.TMNOFLAGS);
+        producer.send(session.createTextMessage(testMessage));
+        xaResource.end(xid, XAResource.TMSUCCESS);
+
+        int prepareOK = xaResource.prepare(xid);
+        Assert.assertEquals(prepareOK, XAResource.XA_OK, "Prepare phase should return XA_OK");
+
+        xaResource.commit(xid, false);
+
+        // Test by consuming the committed message.
+        ConnectionFactory connectionFactory =
+                (ConnectionFactory) initialContext.lookup(ClientHelper.XA_CONNECTION_FACTORY);
+        Connection connection = connectionFactory.createConnection();
+        Session receivingSession = connection.createSession(false, Session.AUTO_ACKNOWLEDGE);
+        MessageConsumer consumer = receivingSession.createConsumer(receivingSession.createQueue(queueName));
+        connection.start();
+        TextMessage message = (TextMessage) consumer.receive(3000);
+
+        Assert.assertNotNull(message, "Didn't receive a message");
+        Assert.assertEquals(message.getText(), testMessage, "Received message should match sent message");
+
+        session.close();
+        xaConnection.close();
+    }
+
+    @Test
+    public void testPublisherWithRollback() throws NamingException, JMSException, XAException, IOException {
+        String queueName = "testPublisherWithRollback";
+        String testMessage = "testPublisherWithRollback-Message";
+        InitialContext initialContext = initialContextBuilder.withXaConnectionFactory()
+                                                             .withQueue(queueName)
+                                                             .build();
+
+        XAConnectionFactory xaConnectionFactory =
+                (XAConnectionFactory) initialContext.lookup(ClientHelper.XA_CONNECTION_FACTORY);
+
+        XAConnection xaConnection = xaConnectionFactory.createXAConnection();
+        XASession xaSession = xaConnection.createXASession();
+        XAResource xaResource = xaSession.getXAResource();
+
+        Session session = xaSession.getSession();
+        Queue queue = session.createQueue(queueName);
+        MessageProducer producer = session.createProducer(queue);
+        xaConnection.start();
+
+        XidImpl xid = new XidImpl(0, "branchId_1".getBytes(), "globalId_1".getBytes());
```

Comment:
Are we using the same XID in every test case? Isn't it better to use an XID generator like we did previously?

Reply:
In every test the same Xid is used. If the id is not cleared after the test duplicate XID is thrown